File: services/25-http4s/src/main/scala/auth/JwtService.scala
package auth

import cats.effect.IO
import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import java.util.Date
import scala.util.Try

// JwtService — signs and verifies JWTs using HS256.
// Secret loaded from JWT_SECRET env var; falls back to "dev-secret".
object JwtService {
  private val secret: String  = sys.env.getOrElse("JWT_SECRET", "dev-secret")
  private val algorithm       = Algorithm.HMAC256(secret)
  // Token valid for 8 hours (28800 seconds).
  private val expiryMs: Long  = 8L * 3600 * 1000

  def sign(userId: String, email: String, name: String): IO[String] = IO {
    JWT.create()
      .withClaim("id", userId)
      .withClaim("email", email)
      .withClaim("name", name)
      .withExpiresAt(new Date(System.currentTimeMillis() + expiryMs))
      .sign(algorithm)
  }

  def verify(token: String): IO[Map[String, String]] = IO.fromTry(Try {
    val decoded = JWT.require(algorithm).build().verify(token)
    Map(
      "id"    -> decoded.getClaim("id").asString(),
      "email" -> decoded.getClaim("email").asString(),
      "name"  -> decoded.getClaim("name").asString(),
    )
  })
}

File: services/api-ktor/src/main/kotlin/com/example/auth/JwtService.kt
package com.example.auth

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.exceptions.JWTVerificationException
import java.util.Date

data class JWTClaims(
    val userId: String,
    val email: String,
    val name: String,
)

object JwtService {

    private val secret: String
        get() = System.getenv("JWT_SECRET") ?: "dev-secret-change-in-prod"

    private val algorithm: Algorithm
        get() = Algorithm.HMAC256(secret)

    fun signToken(userId: String, email: String, name: String): String {
        val now        = System.currentTimeMillis()
        // Session length is set by the active industry profile (HIPAA → 15 min).
        val expiryMs   = com.example.Compliance.sessionTimeoutSeconds * 1000

        return JWT.create()
            .withIssuer("pipeline-studio")
            .withSubject(userId)
            .withClaim("email", email)
            .withClaim("name", name)
            .withIssuedAt(Date(now))
            .withExpiresAt(Date(now + expiryMs))
            .sign(algorithm)
    }

    fun verifyToken(token: String): JWTClaims? {
        return try {
            val verifier = JWT.require(algorithm)
                .withIssuer("pipeline-studio")
                .build()

            val decoded = verifier.verify(token)

            JWTClaims(
                userId = decoded.subject,
                email  = decoded.getClaim("email").asString(),
                name   = decoded.getClaim("name").asString(),
            )
        } catch (e: JWTVerificationException) {
            null
        }
    }
}
